### PR TITLE
Misc Refactoring & Cleanup in libkvs

### PR DESCRIPTION
**Pull request by chu11**
Spliced out patches from a larger RFC11 PR that @garlick and I are working on.  Apologies I missed these and didn't put them in #1152 as I couldn't quite see their independence at first.

All patches are from @garlick.  I've reviewed and it looks good to me.  But as I'm opening this PR, think it's wrong for me to hit the button.

### CODE REVIEW

**Comment by grondo** [comment]
One thing I'd suggest is to add more verbose descriptions in the commit body for these commits.

**Comment by chu11** [comment]
@grondo Agreed, they were a bit too simple.  Updated with more details.

**Comment by chu11** [comment]
Doh, travis caught some warnings that gcc didn't catch for either of our boxes I guess.

```
errno == EINVAL;
```

did get one weird failure

```
FAIL: t0016-cron-faketime.t 5 - flux-cron tab works for Mondays, midnight

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```

I'll assume the error was due to something inside the travis system.  Re-push & try again.

**Comment by chu11** [comment]
Had a couple more "write error" builds, had to restart them.  But got one with:

```FAIL: t0001-basic.t 41 - broker --k-ary option works```

in the raw logs

```
flux-start: warning: setting --bootstrap=selfpmi due to --size option
flux-broker: module 'resource-hwloc' was not cleanly shutdown
flux-broker: module 'aggregator' was not cleanly shutdown
flux-broker: module 'kvs' was not cleanly shutdown
flux-broker: module 'content-sqlite' was not cleanly shutdown
flux-broker: module 'barrier' was not cleanly shutdown
flux-broker: module 'job' was not cleanly shutdown
flux-broker: module 'resource-hwloc' was not cleanly shutdown
flux-broker: module 'kvs' was not cleanly shutdown
flux-broker: module 'aggregator' was not cleanly shutdown
flux-broker: module 'resource-hwloc' was not cleanly shutdown
flux-broker: module 'barrier' was not cleanly shutdown
flux-broker: module 'job' was not cleanly shutdown
flux-broker: module 'aggregator' was not cleanly shutdown
flux-broker: module 'barrier' was not cleanly shutdown
flux-broker: module 'kvs' was not cleanly shutdown
flux-broker: module 'aggregator' was not cleanly shutdown
flux-broker: module 'job' was not cleanly shutdown
flux-broker: module 'resource-hwloc' was not cleanly shutdown
flux-broker: module 'barrier' was not cleanly shutdown
flux-broker: module 'job' was not cleanly shutdown
flux-broker: module 'kvs' was not cleanly shutdown
flux-start: warning: setting --bootstrap=selfpmi due to --size option
flux-broker: module 'kvs' was not cleanly shutdown
flux-broker: module 'content-sqlite' was not cleanly shutdown
flux-broker: module 'aggregator' was not cleanly shutdown
flux-broker: module 'kvs' was not cleanly shutdown
flux-broker: module 'barrier' was not cleanly shutdown
flux-broker: module 'aggregator' was not cleanly shutdown
flux-broker: module 'resource-hwloc' was not cleanly shutdown
flux-broker: module 'resource-hwloc' was not cleanly shutdown
flux-broker: module 'resource-hwloc' was not cleanly shutdown
flux-broker: module 'barrier' was not cleanly shutdown
flux-broker: module 'kvs' was not cleanly shutdown
flux-broker: module 'aggregator' was not cleanly shutdown
flux-broker: module 'aggregator' was not cleanly shutdown
flux-broker: module 'resource-hwloc' was not cleanly shutdown
flux-broker: module 'barrier' was not cleanly shutdown
flux-broker: module 'barrier' was not cleanly shutdown
flux-broker: module 'kvs' was not cleanly shutdown
flux-start: warning: setting --bootstrap=selfpmi due to --size option
flux-broker: module 'resource-hwloc' was not cleanly shutdown
flux-broker: module 'aggregator' was not cleanly shutdown
flux-broker: module 'kvs' was not cleanly shutdown
flux-broker: module 'barrier' was not cleanly shutdown
flux-broker: module 'resource-hwloc' was not cleanly shutdown
flux-broker: module 'aggregator' was not cleanly shutdown
flux-broker: module 'resource-hwloc' was not cleanly shutdown
flux-broker: module 'kvs' was not cleanly shutdown
flux-broker: module 'barrier' was not cleanly shutdown
flux-broker: module 'barrier' was not cleanly shutdown
flux-broker: module 'resource-hwloc' was not cleanly shutdown
flux-broker: module 'kvs' was not cleanly shutdown
flux-broker: module 'aggregator' was not cleanly shutdown
flux-start: 0 (pid 77952) Killed
not ok 41 - broker --k-ary option works
```

Perhaps #1126 ?  Gonna restart it.

**Comment by chu11** [comment]
codecov diff is sort of bad, I"ll try and add some simple unit tests to catch bad input checks tomorrow

File: src/common/libkvs/kvs_commit.c [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    int saved_errno;
+
+    if (!(uuid = zuuid_new ())) {
+        errno = ENOMEM;
+        goto done;
+    }
+    if (!(f = flux_kvs_fence (h, flags, zuuid_str (uuid), 1, txn)))
+        goto done;
+done:
+    saved_errno = errno;
```
Comment (grondo) [inline src/common/libkvs/kvs_commit.c]: Could this function could be simplified (and the confusing `goto done` when the `done:` label is on the next line removed) with something like this?
```C
    if (!(uuid = zuuid_new ()) || !(f = flux_kvs_fence (h, flags, zuuid_str (uuid), 1, txn)))
        saved_errno = errno;
    zuuid_destroy (&uuid);
    errno = saved_errno;
    return f;
```

Also avoids setting errno to assign it to saved_errno, and set it back to errno again.

If taking this approach it might be good to also initialize saved_errno. (then you could also add something like `if (saved_errno) errno = saved_errno;`)

File: src/common/libkvs/kvs_lookup.c [hunk 2]
```diff
@@ -47,6 +68,12 @@ flux_future_t *flux_kvs_lookupat (flux_t *h, int flags, const char *key,
     flux_future_t *f;
     json_t *obj = NULL;
 
+    if (!h || !key || strlen (key) == 0) {
+        errno = EINVAL;
+        return NULL;
+    }
+    if (validate_lookup_flags (flags) < 0)
+        return NULL;
     if (!treeobj) {
```
Comment (grondo) [inline src/common/libkvs/kvs_lookup.c]: just an idle suggestion -- a couple lines of code could be removed by combining the two conditionals here:
```C
    if (!h || !key || strlen (key) == 0 || validate_lookup_flags (flags) < 0) {
        errno = EINVAL;
        return NULL;
    }
```

as a bonus `validate_lookup_flags` doesn't have to set its own errno (it is internal and only fails in one way anyhow)

Comment (chu11) [inline src/common/libkvs/kvs_lookup.c]: i like this idea

File: src/common/libkvs/kvs_commit.c (hunk 1 above)
Comment (chu11) [inline src/common/libkvs/kvs_commit.c]: that would simplify it a bit.

**Comment by grondo** [comment]
Typo in commit message for 1c26ce4? "Simply" should be "Simplify"?

Also, maybe add a descriptive message to 1284050, or perhaps squash it into another commit? (Sorry didn't quite grok if the code was just moved around, in which case it might make more sense to keep your cleanup commit separate). If you feel the cleanup is simple enough to not need a message body, though, I'd be ok with that.

Otherwise, is this ready? Nice work on coverage of added code :+1:

**Comment by chu11** [comment]
@grondo Doh, thanks for the typo catch.

Yeah, not really much to say about https://github.com/flux-framework/flux-core/commit/12840500b33385b6dc40b354cf81d5de66e146b5, as it's just moving code around.  I changed the commit log say "simplify" instead of "cleanup", since that may make the single line make more sense.  I didn't want to squash it into the earlier commit https://github.com/flux-framework/flux-core/pull/1153/commits/de37f360a32d989f82ed1655c0cf84c925150b92, b/c that one was more about splitting the functions out of the ```kvs_txn``` file.

Basically it's all ready for merge.

**Comment by grondo** [comment]
Ok, thanks.

Had to restart one of the builds due to `write error` -- argh that is annoying.
